r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -99,15 +210,15 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         }
         let p_gadget = NonNativeFieldGadget::<SimulationF, ConstraintF> {
             limbs: p_gadget_limbs,
-            num_of_additions_over_normal_form: ConstraintF::one(),
-            is_in_the_normal_form: false,
+            num_of_additions_over_normal_form: BigUint::zero(),
             simulation_phantom: PhantomData,
         };
 
         // Step 2: compute surfeit
-        let surfeit = overhead!(self.prod_of_num_of_additions + ConstraintF::one()) + 1 + 1;
+        let surfeit = ceil_log_2!(BigUint::one() + &self.num_add_over_normal_form);
 
-        // Step 3: allocate k
+        // Step 3: allocate k,
+        // Costs `C = len(p) + surfeit` constraints.
```

Comment:
It costs len(p) + len(num_adds + 1) + 1, which is bigger than len(p) + len(num_adds + 1) as it was claimed in a comment above and it is practically used in the code:
        surfeit = ceil_log_2!(BigUint::one() + &self.num_add_over_normal_form)
WARNING: Recall that the reduced inputs are < 2^len(p) - 1, but they can be > p, that's why we need to add +1 in my opinion.

Reply:
Why it should cost more than `len(p) + ceil_log2(num_adds+1)` constraints? The quotient `k` is strictly bounded by `num_adds +1 ` times `2^len(p)`, otherwise the product `kp` und thus `kp+r` is too large to be reached by a mul result having `num_adds`. Said this, `k` has at most `ceil_log2(num_adds+1) + len(p)` bits.